host/workflowsidinternalratelimit_test.go

```diff
@@ -0,0 +1,230 @@
+		TestClusterConfig:     clusterConfig,
+	}
+	s.IntegrationBase = NewIntegrationBase(params)
+	suite.Run(t, s)
+}
+
+func (s *WorkflowIDInternalRateLimitIntegrationSuite) SetupSuite() {
+	s.setupLogger()
+
+	s.Logger.Info("Running integration test against test cluster")
+	clusterMetadata := NewClusterMetadata(s.T(), s.testClusterConfig)
+	dc := persistence.DynamicConfiguration{
+		EnableCassandraAllConsistencyLevelDelete: dynamicconfig.GetBoolPropertyFn(true),
+		PersistenceSampleLoggingRate:             dynamicconfig.GetIntPropertyFn(100),
+		EnableShardIDMetrics:                     dynamicconfig.GetBoolPropertyFn(true),
+	}
+	params := pt.TestBaseParams{
+		DefaultTestCluster:    s.defaultTestCluster,
+		VisibilityTestCluster: s.visibilityTestCluster,
+		ClusterMetadata:       clusterMetadata,
+		DynamicConfiguration:  dc,
+	}
+	cluster, err := NewCluster(s.T(), s.testClusterConfig, s.Logger, params)
+	s.Require().NoError(err)
+	s.testCluster = cluster
+	s.engine = s.testCluster.GetFrontendClient()
+	s.adminClient = s.testCluster.GetAdminClient()
+
+	s.domainName = s.randomizeStr("integration-test-domain")
+	s.Require().NoError(s.registerDomain(s.domainName, 1, types.ArchivalStatusDisabled, "", types.ArchivalStatusDisabled, ""))
+
+	s.domainCacheRefresh()
+}
+
+func (s *WorkflowIDInternalRateLimitIntegrationSuite) SetupTest() {
+	s.Assertions = require.New(s.T())
+}
+
+func (s *WorkflowIDInternalRateLimitIntegrationSuite) TearDownSuite() {
+	s.tearDownSuite()
+}
+
+func (s *WorkflowIDInternalRateLimitIntegrationSuite) TestWorkflowIDSpecificInternalRateLimits() {
+	const (
+		testWorkflowID   = "integration-workflow-specific-internal-rate-limit-test"
+		testWorkflowType = "integration-workflow-specific-internal-rate-limit-test-type"
+		testTaskListName = "integration-workflow-specific-internal-rate-limit-test-taskList"
+		testIdentity     = "worker1"
+	)
+
+	activityName := "test-activity"
+
+	request := &types.StartWorkflowExecutionRequest{
+		RequestID:                           uuid.New(),
+		Domain:                              s.domainName,
+		WorkflowID:                          testWorkflowID,
+		WorkflowType:                        &types.WorkflowType{Name: testWorkflowType},
+		TaskList:                            &types.TaskList{Name: testTaskListName},
+		Input:                               nil,
+		ExecutionStartToCloseTimeoutSeconds: common.Int32Ptr(100),
+		TaskStartToCloseTimeoutSeconds:      common.Int32Ptr(1),
+		Identity:                            testIdentity,
+		WorkflowIDReusePolicy:               types.WorkflowIDReusePolicyTerminateIfRunning.Ptr(),
+	}
+
+	ctx, cancel := createContext()
+	defer cancel()
+
+	we, err := s.engine.StartWorkflowExecution(ctx, request)
+	s.Nil(err)
+
+	s.Logger.Info("StartWorkflowExecution", tag.WorkflowRunID(we.RunID))
+
+	workflowComplete := false
+	activityCount := int32(5)
+	activityCounter := int32(0)
+
+	dtHandler := func(execution *types.WorkflowExecution, wt *types.WorkflowType,
```

Comment:
Not related to this PR but I think we need a better abstraction to process/complete these workflows so integration code stays more lean and focused to the behavior under test.

Reply:
I completely agree